### DISCUSSION TITLE
Corese and RDF dataset canonicalization

### DISCUSSION THREAD

**Discussion by mereolog** [Wimmics/corese#189] [Q&A] [answered]
https://w3c.github.io/rdf-canon/reports/#subj_Corese_Java has it that Corese fully supports https://www.w3.org/TR/rdf-canon/.

I run a simply test below (using [rdfc10/test053-in.nq](https://w3c.github.io/rdf-canon/tests/rdfc10/test053-in.nq)), but the output file does not correspond to the expected result (e.g., [rdfc10/test053-rdfc10.nq](https://w3c.github.io/rdf-canon/tests/rdfc10/test053-rdfc10.nq)).

I am probably missing something, so any insight from you will be much appreciated.

```
Graph graph = Graph.create();
Load ld = Load.create(graph);
ld.parse("/test053-in.nq");
ResultFormat exporter = ResultFormat.create(graph, ResultFormat.N_QUADS);
String result = exporter.toString();
FileWriter writer = new FileWriter("/test053-out.nq");
writer.write(result);
writer.close();
```

**Comment by remiceres** [comment, marked as answer]
Hello Pawel,

Thank you for your question. I see two things to modify:

First, Canonical-RDF is implemented in the development version of Corese but has not been released yet (it will be released in the next few months with version 4.5.1). You can test it now by downloading the development version (Nightly Build) directly from the release page of Corese: <https://github.com/Wimmics/corese/releases>, file `corese-core-xxxxxxx-nightly.jar`.

Second, you must use the rdfc10 result format `ResultFormat.RDFC10_FORMAT` instead of 'ResultFormat.N_QUADS'. Here is the code with the modification:

```java
Graph graph = Graph.create();

// Load data
Load ld = Load.create(graph);
ld.parse("/test053-in.nq");

// Export data
ResultFormat exporter = ResultFormat.create(graph, ResultFormat.RDFC10_FORMAT);
String result = exporter.toString();

// Write to file
FileWriter writer = new FileWriter("/test053-out.nq");
writer.write(result);
writer.close();
```

If you have any more questions, please don't hesitate to ask.

**Comment by mereolog** [comment]
corese-core-f29b7ae-nightly worked fine for the most of the tests.

But there is a diff for test075-in.nq, where instead of:
```
<http://example.org/vocab#test> <http://example.org/vocab#A> _:c14n0 .
<http://example.org/vocab#test> <http://example.org/vocab#B> _:c14n2 .
_:c14n0 <http://example.org/vocab#next> _:c14n1 .
_:c14n2 <http://example.org/vocab#next> _:c14n1 .
```
I got:
```
<http://example.org/vocab#test> <http://example.org/vocab#A> _:c14n2 .
<http://example.org/vocab#test> <http://example.org/vocab#B> _:c14n0 .
_:c14n0 <http://example.org/vocab#next> _:c14n1 .
_:c14n2 <http://example.org/vocab#next> _:c14n1 .
```

I also got for test060-in.nq:
`ERROR parser.NSManager.toBase:613 [] - Illegal character in opaque part at index 7: urn:ex: `
but this may be due to some encoding issues.

**Reply by remiceres** [reply]
Hello,

Thank you for the report.

For test075, it seems that you are using the classic `RDFC10_FORMAT`, which uses SHA-256. However, this test requires the use of the alternative SHA-384 hash algorithm, `RDFC10_SHA384_FORMAT` as specified in the [W3C RDF Canonicalization Test](https://w3c.github.io/rdf-canon/tests/#test075c).

For test060, there is a warning: `ERROR parser.NSManager.toBase:613 [] - Illegal character in opaque part at index 7: urn:ex:`, and thank you for reporting it. I had overlooked it. However, this does not seem to affect the result of the algorithm, which remains correct. I will look into this issue.

You can find the unit tests I used for canonicalRDF of Corese [here](https://github.com/Wimmics/corese/blob/development/corese-unit-test/src/test/java/fr/inria/corese/w3c/canonicalRdf/canonicalRdfTest.java).
